taichi/program/state_flow_graph.h

```diff
@@ -28,14 +28,42 @@ class StateFlowGraph {
     std::string task_name;
     // Incremental ID to identify the i-th launch of the task.
     int launch_id;
+    bool is_initial_node{false};
+
+    std::unordered_set<AsyncState, AsyncStateHash> input_states, output_states;
 
-    StateToNodeMapping input_edges;
     // Profiling showed horrible performance using std::unordered_multimap (at
     // least on Mac with clang-1103.0.32.62)...
     std::unordered_map<AsyncState, std::unordered_set<Node *>, AsyncStateHash>
-        output_edges;
+        output_edges, input_edges;
 
     std::string string() const;
+
+    bool has_state_flow(AsyncState state, const Node *destination) const {
+      // true: (state flow edge) the state generated by this node is used by the
+      // destination node.
+
+      // false: (dependency edge) the destination node does not
+      // use the generated state, but its execution must happen after this case.
+      // This usually means a write-after-read (WAR) dependency on state.
+
+      // Note:
+      // Read-after-write leads to data flow edges
+      // Write-after-write leads to data flow edges
```

Comment:
I feel like WAW is also a dependency, which prevents overly aggressive reordering? I.e. the second write doesn't use the result of the first write, but it must happen after the first write.

Reply:
Agreed. WAW and WAR can be eliminated by creating another SNode and write to that...

Reply:
...although we don't do that optimization.